torch/_inductor/kernel/mm_common.py

```diff
@@ -421,16 +421,27 @@ def mm_args(
     out_dtype=None,
     use_4x2_dim=False,
     mat2_transposed=False,
+    packed_int4_weights=False,
 ):
     """
     Common arg processing for mm,bmm,addmm,etc
     """
     mat1, mat2 = realize_inputs(mat1, mat2)
     *b1, m, k1 = mat1.get_size()
-    if mat2_transposed:
-        *b2, n, k2 = mat2.get_size()
+    if packed_int4_weights:
+        # weight format is [n/8][k/(InnerKTiles*16)][32][InnerKTiles/2] for int32
```

Comment:
After `_convert_weight_to_int4pack` is called, the weight format is [n/8][k/(InnerKTiles*16)][32][InnerKTiles/2] for int32. Not device-specific.

https://github.com/pytorch/pytorch/pull/129940#issue-2385435984

Reply:
> After `_convert_weight_to_int4pack` is called, the weight format is [n/8][k/(InnerKTiles*16)][32][InnerKTiles/2] for int32. Not device-specific.
> 
> [#129940 (comment)](https://github.com/pytorch/pytorch/pull/129940#issue-2385435984)

A bit confused. Why is a device-agnostic format blocked? How do we decide `InnerKTiles`?

Reply:
`innerKTiles` is decided while quantizing a model.

e.g. https://github.com/pytorch/ao/blob/f5703b07acc683653556d04ef970709ba47dba10/torchao/quantization/GPTQ.py#L662-L667

Reply:
@jgong5 @sanchitintel Sorry, I misunderstand the meaning of `agnostic`. `_convert_weight_to_int4pack` is based on the backends among devices. `[n/8][k/(InnerKTiles*16)][32][InnerKTiles/2]` format is same among the devices.

Reply:
- My understanding is the actual physical layout is backend specific and even instruction related (AVX512 and AVX2). @yanbing-j please correct me if I am wrong
- If so, maybe we should implement a unpack function to unpack back to normal int4 format inside gemm template before we pack it again (AMX might requires a different physical layout). It might be easier to implement the repack function in gemm template.

Reply:
For CPU, the actual physical layout is `[N/64, K, 32]` for `uint8_t`, irrespective of the vectorization ISA used. @yanbing-j, can you please confirm? Thanks!

> before we re-pack it again

@leslie-fang-intel, in the GEMM template, I'm not doing re-packing, but only a view op.
But yes, that may have to be done for AMX in the near future.

Reply:
@leslie-fang-intel I suppose the actual physical layout is backend specific, but they are all same [n/8][k/(InnerKTiles*16)][32][InnerKTiles/2] among devices or AVX2/AVX512.

Reply:
<img width="743" alt="image" src="https://github.com/user-attachments/assets/d24605ee-4363-4273-b5e1-025947545805">